src/python/pants/backend/build_files/fix/deprecations/renamed_fields_rules.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import dataclasses
+import tokenize
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import DefaultDict, Mapping
+
+from pants.backend.build_files.fix.base import FixBuildFilesRequest
+from pants.backend.build_files.fix.deprecations.base import FixBUILDRequest
+from pants.backend.build_files.fix.deprecations.subsystem import BUILDDeprecationsFixer
+from pants.core.goals.fix import FixResult
+from pants.engine.fs import CreateDigest, DigestContents, FileContent
+from pants.engine.internals.native_engine import Digest, Snapshot
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import RegisteredTargetTypes, TargetGenerator
+from pants.engine.unions import UnionMembership
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+
+
+class Request(FixBuildFilesRequest):
+    tool_subsystem = BUILDDeprecationsFixer
+
+
+class RenameRequest(FixBUILDRequest):
+    pass
+
+
+@dataclass(frozen=True)
+class RenamedFieldTypes:
+    """Map deprecated field names to their new name, per target."""
+
+    target_field_renames: FrozenDict[str, FrozenDict[str, str]]
+
+    @classmethod
+    def from_dict(cls, data: Mapping[str, Mapping[str, str]]) -> RenamedFieldTypes:
+        return cls(
+            FrozenDict(
+                {
+                    target_name: FrozenDict(
+                        {
+                            deprecated_field_name: new_field_name
+                            for deprecated_field_name, new_field_name in field_renames.items()
+                        }
+                    )
+                    for target_name, field_renames in data.items()
+                }
+            )
+        )
+
+
+@rule
+def determine_renamed_field_types(
+    target_types: RegisteredTargetTypes, union_membership: UnionMembership
+) -> RenamedFieldTypes:
+    target_field_renames: DefaultDict[str, dict[str, str]] = defaultdict(dict)
+    for tgt in target_types.types:
+        field_types = list(tgt.class_field_types(union_membership))
+        if issubclass(tgt, TargetGenerator):
+            field_types.extend(tgt.moved_fields)
+
+        for field_type in field_types:
+            if field_type.deprecated_alias is not None:
+                target_field_renames[tgt.alias][field_type.deprecated_alias] = field_type.alias
+
+        # Make sure we also update deprecated fields in deprecated targets.
+        if tgt.deprecated_alias is not None:
+            target_field_renames[tgt.deprecated_alias] = target_field_renames[tgt.alias]
+
+    return RenamedFieldTypes.from_dict(target_field_renames)
+
+
+@rule
+def fix_single(
+    request: RenameRequest,
+    renamed_field_types: RenamedFieldTypes,
+) -> FileContent:
```

Comment:
It might be better to have a wrapping result type. It makes the rule graph "purer" and less confusing -- generally, if you want `FileContent`, it's not appropriate to use this rule. This rule is only useful in the extremely specific use case of renaming targets.

Why does it matter? It makes our rule graph error messages slightly less bad, and reduces the risk of weird rule graph issues that end up resulting in super confusing graph issues.